llpc/lower/llpcSpirvLowerGlobal.cpp

```diff
@@ -1916,6 +1917,39 @@ void SpirvLowerGlobal::lowerBufferBlock() {
   }
 }
 
+// =====================================================================================================================
+// Lowers aliased variables.
+void SpirvLowerGlobal::lowerAliasedVal() {
+  std::vector<GlobalVariable *> aliasedVals;
+  unsigned maxInBits = 0;
+  unsigned index = 0;
+  // Aliased variables can contain different byte size, we require the maximum size to be as base variable to replace
+  // the others.
+  for (GlobalVariable &global : m_module->globals()) {
+    auto addrSpace = global.getType()->getAddressSpace();
+    if (addrSpace == SPIRAS_Local) {
+      auto meta = global.getMetadata(gSPIRVMD::Lds);
+      if (!meta)
+        return;
+      const unsigned aliased = mdconst::dyn_extract<ConstantInt>(meta->getOperand(0))->getZExtValue();
+      if (aliased) {
+        unsigned inBits = static_cast<unsigned>(
+            m_module->getDataLayout().getTypeSizeInBits(global.getType()->getPointerElementType()));
+        if (inBits > maxInBits) {
+          maxInBits = inBits;
+          index = aliasedVals.size();
+        }
+        aliasedVals.push_back(&global);
+      }
+    }
+  }
+
+  for (uint32_t i = 0; i < aliasedVals.size(); i++) {
+    if (i != index)
+      replaceGlobal(m_context, aliasedVals[i], aliasedVals[index]);
```

Comment:
I'd like to see some tests to see how this works because it does not seem correct to me.  If you look at the table [here](https://www.khronos.org/registry/SPIR-V/specs/unified1/SPIRV.html#_aliasing), you will notice that two OpVariable in the workgroup cannot be aliased.  It says 

> The Aliased decoration can be used to express that certain [memory object declarations](https://www.khronos.org/registry/SPIR-V/specs/unified1/SPIRV.html#MemoryObjectDeclaration) **may alias**

I'm not sure how replacing one global with another is the correct behaviour.  I was expecting that they would have to have the same layout, but not be the same memory.  Again some test cases could help clearify what is happen, and I would be able to better explain what I think should be happening.

Reply:
I found the spec for the extension, and now I see what is going on.  The change makes sense now.  Can you add a comment referencing the important portions of the spec? This should make it easier to understand.

Reply:
I added a comment and test. Thanks.